pydoctor/sphinx_ext/api_output.py

```diff
@@ -0,0 +1,69 @@
+"""
+Generate the API docs using pydoctor to be integrated into Sphinx build system.
+
+Inside the Sphinx conf.py file you need to define the following configuration options:
+
+* pydoctor_cwd: Absolute path in which the pydoctor build is trigerred.
+* pydoctor_args: A list with all the pydoctor command line arguments used to trigger the build.
+
+You must call pydoctor with `--quiet` argument
+as otherwise any extra output is converted into Sphinx warnings.
+"""
+import os
+from sphinx.util import logging
+from sphinx.errors import ConfigError
+
+from contextlib import redirect_stdout
+from io import StringIO
+from pydoctor import __version__
+from pydoctor.driver import main
+
+logger = logging.getLogger(__name__)
+
+
+def on_build_finished(app, exception):
+    """
+    Called when Sphinx build is done.
+    """
+    if not app.config.pydoctor_args:
+        raise ConfigError("Missing 'pydoctor_args'.")
+
+    if not app.config.pydoctor_cwd:
+        raise ConfigError("Missing 'pydoctor_cwd'.")
+
+    placeholders = {
+        'outdir': app.outdir,
+        }
+
+    args = []
+    for argument in app.config.pydoctor_args:
+        args.append(argument.format(**placeholders))
+
+    stream = StringIO()
```

Comment:
I think it's best if we make sure the stream is closed, by using a `with` block. While the garbage collector will probably do it otherwise, that might not happen immediately.

If I recall correctly, closing the stream does discard the buffered data, so it needs to happen after `getvalue()`.

Reply:
Yep... The buffer is freed https://github.com/python/cpython/blob/4a21e57fe55076c77b0ee454e1994ca544d09dc0/Modules/_io/stringio.c#L567

I will move it into a context manger. Thanks